lang/ruby/lib/avro/json_io.rb

```diff
@@ -0,0 +1,284 @@
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'avro/io'
+require 'json'
+
+module Avro
+  module IO
+    # JsonDecoder keeps track of the current data and reads it using a given data type.
+    #
+    # About validations - If JSONDatumReader has its own read_data implementation, data validations can be moved there instead.
+    # That way reader can reuse Schema.validate method and have same behavior as the writer in case of schema and data mismatch.
+    class JsonDecoder
+      attr_reader :reader
+      attr_accessor :current_data
+
+      def initialize(reader)
+        @reader = reader
+
+        datum = @reader.read
+        datum = '""' if datum.empty?
+        data = JSON.parse(datum, {:quirks_mode => true})
+        @current_data = data
+      end
+
+      def read_null
+        error('null') unless current_data.nil?
+        nil
+      end
+
+      def read_boolean
+        error('boolean') unless Schema.is_boolean?(current_data)
+        current_data
+      end
+
+      def read_int
+        error('"int"') unless Schema.is_integer?(current_data)
+
+        Integer(current_data)
+      end
+
+      def read_long
+        error('"long"') unless Schema.is_long?(current_data)
+
+        Integer(current_data)
+      end
+
+      def read_float
+        error('"float"') unless Schema.is_floating_point?(current_data)
+
+        Float(current_data)
+      end
+
+      def read_double
+        read_float
+      end
+
+      def read_bytes
+        error('"bytes"') unless current_data.is_a?(String)
+        current_data.encode(Encoding::ISO_8859_1, Encoding::UTF_8)
+      end
+
+      def read_fixed(size)
+        error("\"fixed of size: #{size}\"") unless Schema.is_fixed?(current_data, size)
+        read_bytes
+      end
+
+      def read_string
+        error('string') unless current_data.is_a?(String)
+        current_data ? current_data.encode(Encoding::UTF_8) : current_data
+      end
+
+      def error(data_type)
+        raise AvroTypeError.new(data_type, current_data)
+      end
+    end # JsonDecoder
+
+    # Unlike DatumReader, JsonDatumReader reads the file first so that it can decode schema with the JSON data.
+    # It uses JsonDecoder to keep track of current data as it is going through the schema.
+    class JsonDatumReader < DatumReader
+      def read_record(writers_schema, readers_schema, decoder)
+        readers_fields_hash = readers_schema.fields_hash
+        read_record = {}
+
+        data = decoder.current_data
```

Comment:
Since we are reading a record, we assume data is a hash. If not, we should crash which will later when accessing `data[field.name]`. Are you saying we should check and crash beforehand explicitly?